{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "c64bc101-70b2-4ec5-978f-1ed241e40bb9",
   "metadata": {},
   "source": [
    "# SCENIC steps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d001ce85-fae1-4b78-b906-e5aac5643c07",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create basic row and column attributes for the loom file:\n",
    "import loompy as lp\n",
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import glob\n",
    "from cytoolz import compose\n",
    "import numpy as np\n",
    "from pyscenic.transform import df2regulons\n",
    "np.bool = np.bool_\n",
    "import os, glob, re, pickle\n",
    "from pyscenic.aucell import aucell\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from pyscenic.utils import modules_from_adjacencies, load_motifs\n",
    "\n",
    "\n",
    "sc.settings.verbosity = 0 # verbosity: errors (0), warnings (1), info (2), hints (3)\n",
    "#sc.logging.print_versions()\n",
    "sc.set_figure_params(dpi=150, fontsize=10, dpi_save=600)\n",
    "\n",
    "\n",
    "# ranking databases\n",
    "f_db_glob = \"/sc/arion/projects/roussp01a/pengfei/tools/SCENIC/gene_based/*feather\"\n",
    "f_db_names = ' '.join( glob.glob(f_db_glob) )\n",
    "\n",
    "# motif databases\n",
    "f_motif_path = \"/sc/arion/projects/roussp01a/pengfei/tools/SCENIC/motifs-v10nr_clust-nr.hgnc-m0.001-o0.0.tbl\"\n",
    "f_tfs = \"/sc/arion/projects/roussp01a/pengfei/tools/SCENIC/allTFs_hg38.txt\" # human\n",
    "\n",
    "cmc_dir='/sc/arion/projects/CommonMind/roussp01a/ENT/snRNAseq/'\n",
    "ADJACENCIES_FNAME=os.path.join(cmc_dir, 'qc_scanpy/{}_allG.adj.csv'.format('ENTNN'))\n",
    "REGULONS_FNAME=os.path.join(cmc_dir, 'qc_scanpy/{}_allG.reg.csv'.format('ENTNN'))\n",
    "REGULONS_DAT_FNAME = os.path.join(cmc_dir, 'qc_scanpy/{}_allG.regulons.dat'.format('ENTNN'))\n",
    "AUCELL_MTX_FNAME = os.path.join(cmc_dir, 'qc_scanpy/{}_allG.auc.csv'.format('ENTNN'))\n",
    "\n",
    "\n",
    "N_adata_f = sc.read(cmc_dir+'qc_scanpy/ent_nn_merge_rawcount.h5ad')\n",
    "\n",
    "meta_data = pd.read_csv('/sc/arion/projects/roussp01a/liting/Olf/data/ent_nn_merge_cca.metadata.csv', index_col=0)\n",
    "\n",
    "N_adata_f = N_adata_f[meta_data.index]\n",
    "N_adata_f.obs[\"N_types_stage\"]= meta_data.loc[N_adata_f.obs.index,'cca_N_types_stage'] \n",
    "\n",
    "\n",
    "\n",
    "sc.pp.filter_genes(N_adata_f, min_cells=3)\n",
    "\n",
    "df_exp=pd.DataFrame(N_adata_f.X.toarray(), index = N_adata_f.obs_names, columns = N_adata_f.var_names)\n",
    "\n",
    "row_attrs = {\n",
    "    \"Gene\": np.array(N_adata_f.var_names) ,\n",
    "}\n",
    "col_attrs = {\n",
    "    \"CellID\": np.array(N_adata_f.obs_names) ,\n",
    "    \"nGene\": np.array( np.sum(N_adata_f.X.transpose()>0 , axis=0)).flatten() ,\n",
    "    \"nUMI\": np.array( np.sum(N_adata_f.X.transpose() , axis=0)).flatten(),\n",
    "}\n",
    "lp.create( cmc_dir+'ENTNN_n_filtered_scenic_allG.loom', N_adata_f.X.transpose(), row_attrs, col_attrs)\n",
    "f_loom_path_scenic=cmc_dir+'ENTNN_n_filtered_scenic_allG.loom'\n",
    "\n",
    "\n",
    "def derive_regulons(motifs,):\n",
    "    motifs.columns = motifs.columns.droplevel(0)\n",
    "    \n",
    "    def contains(*elems):\n",
    "        def f(context):\n",
    "            return any(elem in context for elem in elems)\n",
    "        return f\n",
    "\n",
    "    # For the creation of regulons we only keep the 10-species databases and the activating modules. We also remove the\n",
    "    # enriched motifs for the modules that were created using the method 'weight>50.0%' (because these modules are not part\n",
    "    # of the default settings of modules_from_adjacencies anymore.\n",
    "    motifs = motifs[np.fromiter(map(contains('activating'), motifs.Context), dtype=np.bool)]\n",
    "\n",
    "    # We build regulons only using enriched motifs with a NES of 3.0 or higher; we take only directly annotated TFs or TF annotated\n",
    "    # for an orthologous gene into account; and we only keep regulons with at least 10 genes.\n",
    "    regulons = list(filter(lambda r: len(r) >= 5, df2regulons(motifs[(motifs['NES'] >= 0.1) \n",
    "                                                                      & ((motifs['Annotation'] == 'gene is directly annotated')\n",
    "                                                                        | (motifs['Annotation'].str.startswith('gene is orthologous to')\n",
    "                                                                           & motifs['Annotation'].str.endswith('which is directly annotated for motif')))\n",
    "                                                                     ])))\n",
    "    \n",
    "    # Rename regulons, i.e. remove suffix.\n",
    "    return list(map(lambda r: r.rename(r.transcription_factor), regulons))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c359c723-2d36-44a2-b32a-bbc35d04700d",
   "metadata": {},
   "outputs": [],
   "source": [
    "for repi in range(1,50):\n",
    "    ## STEP 1: Network inference based on GRNBoost2 from CLI\n",
    "    !/hpc/users/songl05/.conda/envs/Py39_R43_Ju10/bin/pyscenic grn {f_loom_path_scenic} {f_tfs} -o {ADJACENCIES_FNAME} --num_workers 8\n",
    "\n",
    "    ## STEP 2-3: Regulon prediction aka cisTarget from CLI\n",
    "\n",
    "    !/hpc/users/songl05/.conda/envs/Py39_R43_Ju10/bin/pyscenic ctx {ADJACENCIES_FNAME} \\\n",
    "        {f_db_names} \\\n",
    "        --annotations_fname {f_motif_path} \\\n",
    "        --expression_mtx_fname {f_loom_path_scenic} \\\n",
    "        --output {REGULONS_FNAME} \\\n",
    "        --mask_dropouts \\\n",
    "        --num_workers 8 --nes_threshold 2 \n",
    "\n",
    "    from pyscenic.utils import modules_from_adjacencies, load_motifs\n",
    "\n",
    "    ### STEP 4: Cellular enrichment aka AUCell\n",
    "\n",
    "    df_motifs = load_motifs(REGULONS_FNAME)\n",
    "    regulons = derive_regulons(df_motifs)\n",
    "\n",
    "    import warnings\n",
    "    warnings.filterwarnings(\"ignore\")\n",
    "    dfs = []\n",
    "\n",
    "    for i in range(df_motifs.shape[0]):\n",
    "        df=pd.DataFrame(df_motifs[\"TargetGenes\"][i],)\n",
    "        df['tf']=df_motifs[\"TargetGenes\"].index[i][0]\n",
    "        df['MotifID']=df_motifs[\"TargetGenes\"].index[i][1]\n",
    "        dfs.append(df)\n",
    "\n",
    "    tf_target = pd.concat(dfs, ignore_index=True)\n",
    "    tf_target = pd.DataFrame(tf_target)\n",
    "    tf_target = tf_target.rename(columns={'0': 'target','1':'weight'})\n",
    "\n",
    "\n",
    "    tf_target.to_csv(cmc_dir+'entnn_allG_tf_target.csv')\n",
    "\n",
    "    #### AUCell\n",
    "\n",
    "    auc_mtx = aucell(df_exp, regulons, num_workers=26) \n",
    "    #auc_mtx.to_csv(AUCELL_MTX_FNAME) \n",
    "\n",
    "    ### CELL TYPE SPECIFIC REGULATORS - Z-SCORE\n",
    "\n",
    "\n",
    "    signature_column_names = list(auc_mtx.columns)\n",
    "\n",
    "    import pandas as pd\n",
    " \n",
    "\n",
    "    df_scores=pd.concat([auc_mtx, N_adata_f.obs['N_types_stage']],sort=False, axis=1, join='outer' )\n",
    "\n",
    "    df_results = ((df_scores.groupby(by='N_types_stage').mean() - df_scores[signature_column_names].mean())/ df_scores[signature_column_names].std())#.stack().reset_index().rename(columns={'level_1': 'regulon', 0:'Z'})\n",
    "\n",
    "    df_results.to_csv('/sc/arion/projects/roussp01a/liting/Olf/data/scenic_tf_auc/tf_aucscore_rep'+ str(repi) + '.csv')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ae053b54-6f95-41ec-b7fd-a411d0560187",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Py39_R43_Ju10",
   "language": "python",
   "name": "py39_r43_ju10"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
